Add Platform Reliability RFC (#9)
wellcomecollection/docs: rfcs/019-platform_reliability/ci_cd.md

# CI/CD

Continuous Integration / Continuous deployment

## Deployment

###  Continuous deployment

When test pass against a build in a CI environment those changes are deployed to production.

We need:
- To be satisfied that tests in CI ensure that a build is promotable to production.
- To build on the release tool to ensure the deploy step can actually trigger deployments.

> - To connect tests to deployment!

Reviewer: Can we add something here about rolling back, as if we have CD, the confidence in reverting quick is necessary.

> 

Reviewer: ```suggestion
- To be able to roll back to a know working version quickly (quickly to be defined)
```

### Deployment visibility

> We want to be able to see who deployed what & when in the recent past.

Reviewer: ```suggestion
We want to be able to see who recently deployed what & when.
```

> We should provide an authenticated web dashboard showing deployments

Reviewer: ```suggestion
We should provide an authenticated web dashboard showing deployments.

```

We want to know:
- What was deployed, referencing:
  - The PR that was deployed
  - The tests that have been run against the deployed changes (pre/post deployment)
  - The status of a deployment
- When was a change deployed

> 

Reviewer: ```suggestion
- When a change is available but not deployed
```

### Deployment tool 

The release tooling as currently used is too complicated.

> We should have a single step deployment tool that takes a deploy-able artifact from a build and deploys it to an environment.

Reviewer: ```suggestion
We should have a single step deployment tool that takes deploy-able artifact(s) from a build and deploys it to an environment.
```

Updated this, but not sure if it's clear as we need to deploy stacks, which are multiple services, and artifacts.

```
> cd my_project
> wellcome_release_tool deploy build123 staging
Congratulations abc123 (PR123, PR345) deployed to staging!

> > 

Reviewer: ⛵️

```

## Continuous integration

In order to ensure that developers are able to work quickly and effectively together we should:

> - Reduce build times

Reviewer: I've seen it before that tests won't pass if you pass a budget, not sure if this is too dogmatic, but interesting.
